rat/compiler.py

```diff
@@ -1,488 +1,571 @@
+    """
+    if node.arglist is None:
+        return None
+
+    names = []
+    walker = NameWalker()
+    for arg in node.arglist:
+        arg_names = list(walker.walk(arg))
+
+        if len(arg_names) == 1:
+            names.append(arg_names[0])
+        else:
+            # It wasn't possible to get a unique name for each subscript
+            return None
+    return names
+
+
+def get_primary_ast_variable(statement: ast.Statement) -> ast.Variable:
+    """
+    Compute the primary variable reference in a line of code with the rules:
+    1. There can only be one primary variable reference (priming two references to the same variable is still an error)
+    2. If a variable is marked as primary, then it is the primary variable.
+    3. If there is no marked primary variable, then all variables with dataframes are treated as prime.
+    4. If there are no variables with dataframes, the leftmost one is the primary one
+    5. It is an error if no primary variable can be identified
+    """
+
+    @dataclass
+    class PrimaryWalker(RatWalker):
+        marked: ast.Variable = None
+        candidates: List[ast.Variable] = field(default_factory=list)
+
+        def walk_Variable(self, node: ast.Variable):
+            if node.prime:
+                if self.marked is None:
+                    self.marked = node
+                else:
+                    msg = f"Found two marked primary variables {self.marked.name} and {node.name}. There should only" "be one"
+                    raise CompileError(msg, node)
+            else:
+                self.candidates.append(node)
+
+    walker = PrimaryWalker()
+    walker.walk(statement)
+    marked = walker.marked
+    candidates = walker.candidates
+
+    if marked is not None:
+        return marked
+
+    if len(candidates) == 1:
+        return candidates[0]
+
+    if len(candidates) > 1:
+        if len(set(candidate.name for candidate in candidates)) == 1:
+            msg = (
+                f"No marked primary variable but found multiple references to {candidates[0].name}. One reference"
+                " should be marked manually"
+            )
+            raise CompileError(msg, candidates[0])
+        # If all candidate variables don't have subscripts: leftmost variable is primary
+        elif all(not candidate.arglist for candidate in candidates):
+            candidates[0].prime = True
+            return candidates[0]
+        else:
+            msg = (
+                f"No marked primary variable and at least {candidates[0].name} and {candidates[1].name} are"
+                " candidates. A primary variable should be marked manually"
+            )
+            raise CompileError(msg, candidates[0])
+
+    if len(candidates) == 0:
+        msg = f"No primary variable found on line (this means there are no candidate variables)"
+        raise CompileError(msg, statement)
+
+
+@dataclass
+class AugmentDataVariableSubscriptsWalker(RatWalker):
```

Comment:
I've added this walker in the compiler to handle data variables without subscripts. Please read the class docstring.

Reply:
> y ~ normal(mu, sigma);

This makes sense. My assumption at this point was subscripts were always required. (I just hadn't updated all the tests)

I see the index solution. Maybe there's an edge case if we implement scalar data here? I'm not sure. There'd be the other danger that another dataframe can't join in to this by index (and two things named index are not the same).

Reply:
Ah okay. I was going over the tests and some of them weren't equipped with subscripts. Not sure though since adding a `n` column to the dataframe (like in eightschools) pretty does the same thing.

Reply:
I had a closer look at this. I still want to move completely to requiring subscripts.

I like that the index variables are all explicit and I think it makes the models more readable.

For instance:
```
y' ~ normal(mu, sigma);
```

If y and sigma are in a dataframe, you'll potentially get a different sigma for every y.

If just y is in a dataframe, you get one sigma.

I also do not like modifying the ast. The solution there is moving this info into the variable_table.

I'm gonna leave this for now and merge stuff since tests are passing at all. Any complaints if we remove this in future tho?

Reply:
Ohhh shoot yeah ambiguous operations is a big nono.

I agree on going with the all subscripted variables rule. Right now there's just too much additional logic required even to let the user omit the simplest ones.

Reply:
> If y and sigma are in a dataframe, you'll potentially get a different sigma for every y.

If just y is in a dataframe, you get one sigma.

On second thought, I actually think this should be the right behavior. If sigma is in your dataframe and youre invoking it, then it becomes data. If not then its a parameter. Remember we always see rat models within the context of the supplied dataframes.

I'm not saying the onus of making well sefined code is upon the user, but if sigma was the observed errors I would've written it as `sigma_obs`.

This is a design decision and I'm still very much fine with requiring subscripts though.